### PR TITLE
Track memory attributes to avoid memory leak

### PR DESCRIPTION
**Pull request by mschneid**
Motivation:
In commit c354fa48e10de847cf17c10083a2cbc2c0a63a36 (https://github.com/netty/netty/pull/10209) the way attributes are decoded has changed. In `HttpPostStandardRequestDecoder#decodeAttribute(ByteBuf, Charset)` a buffer is allocated but may not be released while calling `InterfaceHttpPostRequestDecoder#destroy()`.

```
HttpPostRequestDecoder decoder =
    new HttpPostRequestDecoder(new MemoryHttpDataFactory(false), request);
try {
    ...
} finally {
    decoder.destroy();
}
```

`decoder.destroy()` invokes `DefaultHttpDataFactory#cleanRequestHttpData(HttpRequest)`, but as the memory attributes are not tracked, nothing is released.

Workaround:
```
HttpPostRequestDecoder decoder =
    new HttpPostRequestDecoder(new MemoryHttpDataFactory(false), request);
try {
    ...
} finally {
    decoder.getBodyHttpDatas()
        .forEach(ReferenceCounted::release);
    decoder.destroy();
}
```

Modification:
Track memory attributes so that they are released by calling `HttpDataFactory#cleanAllHttpData()` or `HttpDataFactory#cleanRequestHttpData(HttpRequest)`.

Result:
Avoid memory leak

### CODE REVIEW

**Comment by normanmaurer** [comment]
@mschneid can you add a unit test as well ?

**Comment by mschneid** [comment]
> @mschneid can you add a unit test as well ?

@normanmaurer Done

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@mschneid there are unit test failures. PTAL

**Comment by mschneid** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by mschneid** [comment]
@normanmaurer should I squash this PR?

**Comment by normanmaurer** [comment]
@mschneid let me have a look first. From a quick look I am not sure if this PR is 100 % correct. I will come back to you shortly

**Comment by normanmaurer** [comment]
@mschneid so I did spent some time to look into this and the commit 9a701d523193828c180cdfff30bfe803d9b61362 seems to be not correct. Let me think about this a bit and come up with an alternative way of fixing the test-failures

**Comment by jameskleeh** [comment]
It was decided this is a breaking change in my previous PR https://github.com/netty/netty/pull/9672

**Comment by mschneid** [comment]
Any news on that?

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DefaultHttpDataFactoryTest.java [hunk 1]
```diff
@@ -162,4 +162,85 @@ public void removeHttpDataFromCleanShouldIdentifiesDataByTheirIdentities() throw
...
     }
+
+    @Test
+    public void cleanAllHttpDataShouldReleaseAllDiskAttributes() throws Exception {
+        cleanAllHttpDataShouldReleaseAllAttributes(new DefaultHttpDataFactory());
+    }
+
+    @Test
+    public void cleanRequestHttpDataShouldReleaseAllDiskAttributes() throws Exception {
+        cleanRequestHttpDataShouldReleaseAllAttributes(new DefaultHttpDataFactory());
```
Comment (xiexq2019) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DefaultHttpDataFactoryTest.java:173]: test

**Comment by mschneid** [comment]
This PR can be closed because the problem has been fixed with #11188.

**Pull request closed without merging**